doc/api/vm.md

```diff
@@ -1616,6 +1614,156 @@ inside a `vm.Context`, functions passed to them will be added to global queues,
 which are shared by all contexts. Therefore, callbacks passed to those functions
 are not controllable through the timeout either.
 
+## Support of dynamic `import()` in compilation APIs
+
+The following APIs support an `importModuleDynamically` option to enable dynamic
+`import()` in code compiled by the vm module.
+
+* `new vm.Script`
+* `vm.compileFunction()`
+* `new vm.SourceTextModule`
+* `vm.runInThisContext()`
+* `vm.runInContext()`
+* `vm.runInNewContext()`
+* `vm.createContext()`
+
+This option is still part of the experimental modules API. We do not recommend
+using it in a production environment.
+
+### When the `importModuleDynamically` option is not specified or undefined
+
+If this option is not specified, or if it's undefined, code containing
+`import()` can still be compiled by the vm APIs, but when the compiled code is
+executed and it actually calls `import()`, the result will reject with
+[`ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING`][].
+
+### When `importModuleDynamically` is `vm.constants.USE_MAIN_CONTEXT_DEFAULT_LOADER`
+
+This option is currently not supported for `vm.SourceTextModule`.
+
+With this option, when an `import()` is initiated in the compiled code, Node.js
+would use the default ESM loader from the main context to load the requested
+module and return it to the code being executed.
+
+```mjs
+import { Script, constants } from 'node:vm';
+import { resolve } from 'node:path';
+import { writeFileSync } from 'node:fs';
+
+const data = JSON.stringify({ hello: 'world' });
+writeFileSync(resolve(import.meta.dirname, 'foo.json'), data);
+
+const script = new Script('import("./foo.json", { with: { type: "json" } })', {
+  filename: resolve(import.meta.dirname, 'test-with-default.js'),
+  importModuleDynamically: constants.USE_MAIN_CONTEXT_DEFAULT_LOADER,
+});
```

Comment:
This seems like it’s a CommonJS script? But it’s a bit of an awkward example since in CommonJS most users would use `require` to get a JSON file, rather than dynamic `import`. Maybe change it to a dynamic import of an ES module, with a comment above like `// Import ES module into CommonJS code`?

I assume we can’t show how this would be used with an ESM script because this doesn’t yet support `SourceTextModule`?

Reply:
This is not a CommonJS script, it's just a script. If you run it in a new context, there is no `require()` at all (there also isn't going to be process, or URL, or anything not implemented by V8).